torch/csrc/throughput_benchmark/ThroughputBenchmark.cpp

```diff
@@ -0,0 +1,125 @@
+  inputs_.push_back(std::move(input));
+}
+std::shared_ptr<jit::script::Module> ThroughputBenchmark::getModule() {
+  return module_;
+}
+
+at::IValue ThroughputBenchmark::runOnce(int idx) const {
+  TORCH_CHECK(inputs_.size() > idx, inputs_.size(), idx);
+  return module_->forward(inputs_[idx]);
+}
+
+ThroughputBenchmark::ThroughputBenchmark(
+    std::shared_ptr<jit::script::Module> module,
+    BenchmarkConfig config)
+    : module_(module), config_(config) {}
+
+ExecutionStats ThroughputBenchmark::benchmark() const {
+  TORCH_CHECK(
+      config_.num_worker_threads() == 1,
+      "Only parallelization by callers is supported");
+
+  // We pre-generate inputs for each of the threads and avoid threadsafety
+  // issues for random calls. As thread safe random is not provided in the
+  // standard library. We also replicate IValues. So we don't need to copy
+  // them during the main benchmark phase.
+  std::vector<std::vector<std::vector<at::IValue>>> thread_inputs(
+      config_.num_calling_threads());
+  std::vector<std::vector<std::vector<at::IValue>>::iterator> input_iters(
+      config_.num_calling_threads());
+  {
+    std::random_device seeder;
+    std::mt19937 engine(seeder());
+    TORCH_CHECK(!inputs_.empty(), inputs_.size());
+    std::uniform_int_distribution<int> dist(0, inputs_.size() - 1);
+
+    for (int thread_id = 0; thread_id < config_.num_calling_threads();
+         ++thread_id) {
+      // Just in case we generate num_iters inputs for each of the threads
+      // This was if one thread does all the work we will be fine
+      for (int i = 0; i < config_.num_iters() + config_.num_warmup_iters();
+           ++i) {
+        thread_inputs[thread_id].push_back(inputs_[dist(engine)]);
+      }
+      input_iters[thread_id] = thread_inputs[thread_id].begin();
+    }
+  }
+
+  std::mutex m;
+  std::condition_variable cv;
+  std::atomic<int64_t> initialized{0};
+  std::atomic<bool> shutdown{false};
+  std::atomic<bool> start{false};
+  std::atomic<int64_t> num_forwards{0};
+  std::vector<std::thread> callers;
+
+  for (auto thread_id = 0; thread_id < config_.num_calling_threads();
+       ++thread_id) {
+    callers.emplace_back([&, thread_id]() {
+      // We use conditional variable as a barrier to make sure each thread
+      // performs required warmeup iterations before we start measuring
+      for (auto j = 0; j < config_.num_warmup_iters(); ++j) {
+        module_->forward(std::move(*input_iters[thread_id]));
+        ++input_iters[thread_id];
+      }
+      LOG(INFO) << "Initialized thread " << thread_id;
+      ++initialized;
+      cv.notify_all();
+      {
+        std::unique_lock<std::mutex> g(m);
+        cv.wait(g, [&]() -> bool { return start; });
+      }
+      LOG(INFO) << "Starting forward thread " << thread_id;
+      while (!shutdown) {
+        VLOG(2) << "Starting forward pass on thread " << thread_id;
+        module_->forward(std::move(*input_iters[thread_id]));
+        ++input_iters[thread_id];
```

Comment:
it can potentially go out of bounds - there's no guarantee that the main thread will stop things in time

Reply:
oops. Yeah, probably just better to make number of total iterations precise